Mntor 1168/toast notification

toufali opened the pull request:
<!-- The following is intended to be helpful to you. Feel free to remove anything that is not. -->

# References: 
Jira: MNTOR-1168
Figma: https://www.figma.com/file/gvyyqS9egsGjsH0LUDqT1m/Breach-Resolution-Flow-Design?node-id=643%3A21602&t=M7gAMoMdvdtaa8yg-0


<!-- When adding a new feature: -->

# Description
Adds a site-wide "toast" component to display errors or alerts in response to user action

To-do under separate ticket (MNTOR-1271):
- audit the codebase to see where we might display these alerts/errors
- localize error messages

# Screenshot (if applicable)
<img width="1056" alt="Screen Shot 2023-02-24 at 4 03 34 PM" src="https://user-images.githubusercontent.com/14863500/221323575-cee14fda-5b8b-47b6-92ee-49b4a9689996.png">



# How to test
1. Visit the Settings page
2. Add a new, unverified email
3. Trigger a "Too many requests" error by clicking on "Resend verification email" for the new email


# Checklist (Definition of Done)
- [ ] Localization strings (if needed) have been added.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [x] I've added or updated the relevant sections in readme and/or code comments
- [ ] I've added a unit test to test for potential regressions of this bug.
- [ ] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [ ] All acceptance criteria are met.
- [x] Jira ticket has been updated (if needed) to match changes made during the development process.
- [x] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

Vinnl commented on src/client/js/components/toast-alert.js:
```diff
@@ -0,0 +1,133 @@
+    transform-origin: top;
+    transition: transform 300ms;
+    animation: fly-in 300ms, fade-out ${fadeDuration}ms ${ttl}ms forwards;
+    z-index: 2;
```
Can you leave a comment stating what elements we're overriding here?

Vinnl commented on src/client/js/components/toast-alert.js:
```diff
@@ -0,0 +1,133 @@
+    }
+  }
+</style>
+<output><slot></slot><button>✕</button></output>
```
Could you add an aria label for the button?

Vinnl commented on src/client/js/components/toast-alert.js:
```diff
@@ -0,0 +1,133 @@
+    const toasts = Array.from(document.querySelectorAll('toast-alert')).reverse()
+
+    for (let i = 1, y = 0; i < toasts.length; i++) {
+      // aggragate toast heights to push old toasts down as needed
```
```suggestion
      // aggregate toast heights to push old toasts down as needed
```

Mad props to Matt's eagle eyes

Vinnl commented on src/client/js/components/toast-alert.js:
```diff
@@ -0,0 +1,133 @@
+  connectedCallback () {
+    const toasts = Array.from(document.querySelectorAll('toast-alert')).reverse()
+
+    for (let i = 1, y = 0; i < toasts.length; i++) {
```
Could you add a small comment explaining this loop?

Vinnl reviewed (approved):
Looks good. Some observations:

- there's no hover state for the close button
- the capitalisation is inconsistent: "Re-sending verification email failed: Too Many Requests"
- the notification automatically disappears - could be nice to have it stay on screen if the user's hovering it, for example
- the page scrolls up when showing the notification
- I can't reach the notification or close it via keyboard

toufali commented:
Thanks for your review @Vinnl !

I address most comments – Follow-up notes:

>     there's no hover state for the close button
added via https://github.com/mozilla/blurts-server/pull/2868/commits/04fbddab43ca21930375b2e53de008e3efc44de7

>     the capitalisation is inconsistent: "Re-sending verification email failed: Too Many Requests"
This is just using the error message sent from server.  We can optimize this in the follow-up ticket mentioned in description 

>     the notification automatically disappears - could be nice to have it stay on screen if the user's hovering it, for example
added via https://github.com/mozilla/blurts-server/pull/2868/commits/91e01bfaedcc0eaa09a7edd8fe16298d1963981f

>     the page scrolls up when showing the notification
I triggered a toast on the Breaches page and the behavior did not occur.  I believe this behavior is specific to the Settings page – If I recall they do a page reload (which we should improve with a future ticket)

>     I can't reach the notification or close it via keyboard
I'm not sure I would expect to close a toast message escape or similar.  I think since the toast auto-hides, this functionality might not be necessary.  Let me know if you feel strongly otherwise.
